runtime/src/main/java/org/corfudb/runtime/collections/ProtobufIndexer.java

```diff
@@ -23,43 +26,136 @@
                         (key, val) -> ClassUtils.cast(val.getPayload().getField(fieldDescriptor)));
     }
 
-    void registerIndices(final Descriptors.FieldDescriptor fieldDescriptor) {
+    private <T> Index.Spec<Message, CorfuRecord<Message, Message>, ?>
+    getComplexIndex(String indexName) {
+        return new Index.Spec<>(
+                () -> indexName,
+                (Index.Function<Message, CorfuRecord<Message, Message>, T>)
+                        (key, val) -> {
+                            // Get all nested fields, '.' separated
+                            String[] nestedFields = indexName.split("\\.");
+
+                            if (nestedFields.length > 0) {
+                                Message subMessage = val.getPayload();
+
+                                for (int i =0; i < nestedFields.length; i++) {
+                                    try {
+                                        subMessage = (Message) subMessage.getField(
+                                                subMessage.getDescriptorForType().findFieldByName(nestedFields[i]));
+                                    } catch (ClassCastException ce) {
+                                        // Valid cast exception if its the last level (primitive type), but if the '.' separated secondary
+                                        // key has still remaining levels, it is not well-formed.
+                                        if (i != (nestedFields.length - 1)) {
+                                            throw new IllegalArgumentException("Malformed secondary key=" + indexName + " primitive field :: " + nestedFields[i]);
+                                        }
 
+                                        return ClassUtils.cast(subMessage.getField(
+                                                subMessage.getDescriptorForType().findFieldByName(nestedFields[i])));
+                                    }
+                                }
+
+                                return ClassUtils.cast(subMessage);
+                            }
+                            return null;
+                        });
+    }
+
+    void registerIndices(final Descriptors.FieldDescriptor fieldDescriptor, Message payloadSchema) {
         if (fieldDescriptor.getOptions().getExtension(CorfuOptions.schema).getSecondaryKey()) {
             final String indexName = fieldDescriptor.getName();
             if (fieldDescriptor.getType() == FieldDescriptor.Type.GROUP) {
                 throw new IllegalArgumentException("group is a deprecated, unsupported type");
             }
             indices.put(indexName, getIndex(indexName, fieldDescriptor));
+        } else if (fieldDescriptor.getOptions().getExtension(CorfuOptions.schema).hasNestedSecondaryKeys()) {
+            if (fieldDescriptor.getType() == FieldDescriptor.Type.GROUP) {
+                throw new IllegalArgumentException("group is a deprecated, unsupported type");
+            }
```

Comment:
Shouldn't we also handle else if (fieldDescriptor.isRepeated()) { for indexes on collections?

Reply:
I didn't consider collections, let me work on adding it.

Reply:
Done, now it supports collections and nested secondary indexes.